src/server/dflycmd.cc

```diff
@@ -210,30 +297,122 @@ void DflyCmd::Expire(CmdArgList args, ConnectionContext* cntx) {
   return rb->SendOk();
 }
 
+OpStatus DflyCmd::StartFullSyncInThread(FlowInfo* flow, EngineShard* shard) {
+  DCHECK(!flow->fb.joinable());
+
+  SaveMode save_mode = shard == nullptr ? SaveMode::SUMMARY : SaveMode::SINGLE_SHARD;
+  flow->saver.reset(new RdbSaver(flow->conn->socket(), save_mode, false));
+
+  if (shard != nullptr) {
+    flow->saver->StartSnapshotInShard(false, shard);
+  }
+
+  flow->fb = ::boost::fibers::fiber(&DflyCmd::FullSyncFb, this, flow);
+  return OpStatus::OK;
+}
+
+void DflyCmd::FullSyncFb(FlowInfo* flow) {
+  error_code ec;
+  RdbSaver* saver = flow->saver.get();
+
+  if (saver->Mode() == SaveMode::SUMMARY) {
+    auto scripts = sf_->script_mgr()->GetLuaScripts();
+    saver->SaveHeader(scripts);
+  } else {
+    saver->SaveHeader({});
+  }
+
+  if (ec) {
+    LOG(ERROR) << ec;
+    return;
+  }
+
+  if (saver->Mode() != SaveMode::SUMMARY) {
+    // TODO: we should be able to stop earlier if requested.
+    ec = saver->SaveBody(nullptr);
+    if (ec) {
+      LOG(ERROR) << ec;
+      return;
+    }
+  }
+
+  ec = flow->conn->socket()->Write(io::Buffer(flow->eof_token));
+  if (ec) {
+    LOG(ERROR) << ec;
+    return;
+  }
+
+  ec = flow->conn->socket()->Shutdown(SHUT_RDWR);
+}
+
 uint32_t DflyCmd::CreateSyncSession() {
   unique_lock lk(mu_);
 
-  auto [it, inserted] = sync_infos_.emplace(next_sync_id_, new SyncInfo);
+  auto sync_info = make_shared<SyncInfo>();
+  sync_info->flows.resize(shard_set->size() + 1);
+
+  auto [it, inserted] = sync_infos_.emplace(next_sync_id_, std::move(sync_info));
   CHECK(inserted);
 
   return next_sync_id_++;
 }
 
+void DflyCmd::UnregisterFlow(FlowInfo* flow) {
+  flow->conn = nullptr;
+  flow->saver.reset();
+}
+
 void DflyCmd::DeleteSyncSession(uint32_t sync_id) {
+  shared_ptr<SyncInfo> sync_info;
+
+  // Remove sync_info from map.
+  // Store by value to keep alive.
+  {
+    unique_lock lk(mu_);
+
+    auto it = sync_infos_.find(sync_id);
+    if (it == sync_infos_.end())
+      return;
+
+    sync_info = it->second;
+    sync_infos_.erase(it);
+  }
+
+  // Wait for all operations to finish.
+  // Set state to CANCELLED so no other operations will run.
+  {
+    unique_lock lk(sync_info->mu);
+    sync_info->state = SyncState::CANCELLED;
+  }
```

Comment:
See my first comment. We make sure no operations are running and transition into the `CANCELLED` state. In my opinion this is better than a full block because it allows other operations to cancel early on

Reply:
sounds good.